Crossgen2 ARM64 runs & initial cross-targeting support

trylek opened the pull request:
This change introduces initial provisions for dynamically loading
the native JIT based on the targeting OS and architecture; the change
expects the potentially multiple versions of clrjit to coexist with
Crossgen2 in the same folder, marked by a OS / arch suffix. Based
on these prerequisites the change adds ARM64 jobs to the Crossgen2
pipeline.

Thanks

Tomas

trylek committed:
Crossgen2 ARM64 runs

trylek committed:
Enable framework crossgenning on ARM64

trylek committed:
Fix cross-compilation native bin folder for arm/arm64

trylek committed:
Revert incorrect attempt at path fix

trylek committed:
Add jitinterface to crosscomponents.cmake; crossgen2 cross-package

trylek committed:
Add missing targetarch option to Crossgen2 framework compilation

trylek committed:
Initial support for Crossgen2 multitargeting

This change introduces initial provisions for dynamically loading
native Crossgen2 components (jitinterface, clrjit) based on the
hosting / targeting OS and architecture. In general, jitinterface
depends on the hosting OS and architecture, clrjit depends on both
host and target characteristics.

As of this commit, the host environment is represented by the first
subfolder level beneath the "crossgen2" executable; its currently
supported name combinations are host-win-x64, host-win-arm64,
host-lnx-x64, host-lnx-arm64. Targeting environment is represented
by the second-level subfolder under the hosting folder; its currently
supported name combinations are target-win-x64, target-win-arm64,
target-lnx-x64, target-lnx-arm64.

Thanks

Tomas

trylek committed:
Fix one overlooked NativeLibraryPrefix

trylek committed:
Fix race condition in clrjit loading

AntonLapounov commented on src/coreclr/build-test.cmd:
```diff
@@ -646,6 +642,7 @@ if defined __CompositeBuildMode (
     mkdir !__CompositeOutputDir!
     echo --composite>>!__CompositeResponseFile!
     echo -O>>!__CompositeResponseFile!
+    echo --targetarch:%__BuildArch%>>!__CompositeResponseFile!
```
What happens if `!__CompositeOutputDir!` already exists?  Will we append to the already existing `!__CompositeResponseFile!` file?  We might want to `rmdir /s /q !__CompositeOutputDir!` first.

Also I noticed we are using `composite.out` directory name on Windows, but not on Linux.  Could we behave the same way on both OSes?

trylek replied:
For now I have added a command to delete the response file upfront. For other changes related to framework crossgenning, I would prefer to leave them for a separate PR if you don't mind as I already have a half-working private change switching over framework crossgenning to use R2RTest which would substantially simplify both versions of the build-test script and make many of these concerns moot.

trylek committed:
Fix target architecture and callin convention

trylek committed:
Fix bug in targetarch specification in CLRTest.Crossgen.targets

trylek committed:
Fix typo in 'del' command syntax

trylek committed:
Initialize JIT library and interface on the main thread

trylek commented:
To my own surprise I see only one error on ARM64, the GitHub_27924 JitBlue regression test. I'll create a GitHub tracking issue and I'll push another commit adding it to issues.targets.

trylek commented:
OK, the issue on Windows is apparently https://github.com/dotnet/runtime/issues/34316.

trylek commented on src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.cs:
```diff
@@ -58,16 +58,140 @@ private enum ImageFileMachine
+
+            string crossgenPath = Path.GetDirectoryName(typeof(CorInfoImpl).Assembly.Location);
+            string jitHostPath = Path.Combine(crossgenPath, GetHostSpec());
+            string jitHostLibraryName = Path.Combine(jitHostPath, GetLibraryPrefix() + JitSupportLibrary + GetLibraryExtension());
```
Without switching jitinterface we would need to publish two versions of Crossgen2 for arm64, one cross-compiling (x64-hosted) for compiling the framework on the x64 build machine and another native arm64-hosted jitinterface used when running Crossgen2 on the ARM64 Helix HW to compile the test apps.

trylek replied:
In this sense the jitinterface is dependent on the host whereas clrjit is dependent on both the host and target, that's the gist of this change. I believe it smoothly translates to supplying additional variants of jitinterface / jit to achieve all cross-targeting we need.

trylek replied:
For instance, we can add x86 support tomorrow by just publishing x86 jitinterface and clrjit along with Crossgen2 to the proper folders and fixing 2-3 bugs.

trylek replied:
And at some point we can end up merging them all by just creating a union of the subfolders and getting a fully cross-targeting Crossgen2 package.

trylek replied:
Hmm, not sure about x86, we'd probably need Crossgen2 with the 32-bit set but anyway :-).

davidwrighton replied:
@trylek that only makes sense if the crossgen2 compiler package we produce is platform agnostic. I would expect us to configure the world such that the x64 crossgen2 can target arm64 and x64.

MichalStrehovsky replied:
Yup, what David says - crossgen2 is not an AnyCpu project - it's compiled for a specific host OS/arch. This host OS/arch will match the host OS/arch of the SDK itself, the host OS/arch of JitSupport, and the host OS/arch of clrjit.

We need a probing logic for clrjit, because because we have multiple clrjits per __target__ (but still there is just one __host__).

JitSupport is __target__ agnostic.

trylek replied:
OK, I treat this as a preliminary consensus that CoreCLR arm64 build should end up building two Crossgen2 packages - one cross-targeting under <code>artifacts\bin\coreclr\Windows_NT.arm64.Checked\x64\crossgen2</code> that's going to be used for cross-compilation of the framework assemblies on the build machines and a native arm64-hosted under <code>artifacts\bin\coreclr\Windows_NT.arm64.Checked\crossgen2</code> and <code>CORE_ROOT\crossgen2</code> that's going to be subsequently used for native arm64 test compilation on the Helix arm64 machines. I'm going to modify (simplify) the change based on your suggestion.

trylek committed:
Block out the Dev_27924 test on CG2 ARM64 with issue #34316

AntonLapounov commented on src/coreclr/tests/issues.targets:
```diff
@@ -939,6 +939,13 @@
         </ExcludeList>
     </ItemGroup>
 
+    <!-- Crossgen2 on ARM64 specific -->
+    <ItemGroup Condition="'$(XunitTestBinBase)' != '' and '$(TestBuildMode)' == 'crossgen2' and '$(RuntimeFlavor)' == 'coreclr' and ('$(TargetArchitecture)' == 'arm64' or '$(AltJitArch)' == 'arm64')">
+        <ExcludeList Include="$(XunitTestBinBase)/JIT/Regression/JitBlue/GitHub_27924/GitHub_27924/*">
+            <Issue>https://github.com/dotnet/runtime/issues/34316</Issue>
```
According to the issue, it fails on x64 as well.  Why don't we add this exclusion to the previous `<ItemGroup>` then?

trylek replied:
Well, I only saw it on arm64 and I haven't found it in issues.targets. I'll move it to the general Crossgen2 group.

trylek replied:
Fixed in 15th commit. I have also sorted the Crossgen2 issues.targets items alphabetically for easier orientation.

AntonLapounov commented on eng/pipelines/coreclr/crossgen2.yml:
```diff
@@ -22,9 +22,10 @@ jobs:
+    - Linux_arm64
     - OSX_x64
     - Windows_NT_x64
-    - CoreClrTestBuildHost # Either OSX_x64 or Linux_x64
+    - Windows_NT_arm64
```
Bad merge?

trylek replied:
Well, I originally deleted it intentionally as I saw both platforms were already in the list but you're right, I shouldn't be deleting this as at some point it may get switched over to something completely different. Thanks for pointing that out!

trylek replied:
Restored in 15th commit.

trylek committed:
Address Anton's PR feedback

trylek committed:
Revert dynamic loading of jitinterface per PR feedback

trylek committed:
Fix Crossgen2 file list in the installer to match the new structure

trylek committed:
One more fix for the new Crossgen2 package structure

jkotas commented on src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.cs:
```diff
@@ -110,14 +114,68 @@ private IntPtr AllocException(Exception ex)
+            string targetOSComponent = targetOS switch
+            {
+                TargetOS.Windows => "win",
+                TargetOS.Linux => "lnx",
```
Nit: I would use `linux` same as what we are using e.g. for RIDs. I do not think we are using "lnx" as the abbreviation for Linux anywhere else.

trylek replied:
Fixed in 19th commit, thanks for pointing that out.

trylek committed:
Rename OS target component from 'lnx' to 'linux' per JanK's feedback

jkotas commented on src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.cs:
```diff
@@ -110,14 +114,68 @@ private IntPtr AllocException(Exception ex)
+            Debug.Assert(s_jitLibrary == IntPtr.Zero);
+
+            string crossgenPath = Path.GetDirectoryName(typeof(CorInfoImpl).Assembly.Location);
+            string jitTargetPath = Path.Combine(crossgenPath, GetTargetSpec(targetOS, targetArchitecture));
```
We are renaming the JIT .dll anyway (from clrjit to clrjitilc). Would it be simpler to have the target arch in the name, so that we do not have to deal with loading it from subdirectories?

E.g. if the JIT dll was next to `typeof(CorInfoImpl).Assembly.Location`, you can just use the longer overload of `NativeLibrary.Load` to load it and it will take care of the prefixes and suffixes for you.

trylek replied:
I agree it's probably easier for general maintenance, will do. (With the downside that I'll need to fork the call to NativeLibrary.Load for the "default case" vs. "explicitly specified jitPath".)

jkotas replied:
> fork the call to NativeLibrary.Load for the "default case" vs. "explicitly specified jitPath".

Sounds good.

trylek replied:
Fixed in 21st commit.

jkotas commented on src/coreclr/src/tools/Common/JitInterface/JitConfigProvider.cs:
```diff
@@ -29,7 +30,12 @@ public static JitConfigProvider Instance
+            TargetArchitecture targetArchitecture,
+            IEnumerable<CorJitFlag> jitFlags,
+            IEnumerable<KeyValuePair<string, string>> parameters,
+            string jitPath = null)
```
Now that you are loading the JIT .dll manually, you should not need the SetDllImportResolver below.

Also, the jitPath won't work anymore with these changes. Should we keep it working?

trylek replied:
Well, I guess it's probably no big deal to fix CorInfoImpl.Startup to optionally accept the jitPath and use it to override the synthesized subfolder when present. I'll get rid of the SetDllImportResolver based on your suggestion.

trylek replied:
Cleaned up in 20th commit.

jkotas replied:
The jit path is filename (not directory name) of the .dll to load today. 

The jit path exists for convenience so that you do not have to copy things around if you happen to have JIT .dll someplace. I think it would be best to keep the jit path as a simple filename (no additional substitutions/mangling).

trylek replied:
Thanks for clarifying, that makes sense to me and I believe this is exactly how I modified the code in 20th commit.

jkotas replied:
👍

trylek committed:
Add explicit handling of jitPath to Startup; remove SetDllImportResolver

jkotas commented on src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.cs:
```diff
@@ -110,14 +114,76 @@ private IntPtr AllocException(Exception ex)
+            string targetOSComponent = targetOS switch
+            {
+                TargetOS.Windows => "win",
+                TargetOS.Linux => "linux",
```
The JIT .dll is same for `linux`, `osx` and all other Unix systems. Should we have fold linux, osx cases (and all other Unix cases) together into a single name?

trylek replied:
Good point, will rename to "unix".

trylek replied:
Fixed in 21st commit.

trylek committed:
Remove clrjit subfolder, suffix its name with target OS / arch

jkotas commented on src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.cs:
```diff
@@ -110,14 +114,67 @@ private IntPtr AllocException(Exception ex)
+            s_jitStartup(GetJitHost(JitConfigProvider.Instance.UnmanagedInstance));
+        }
+
+        private static string GetLibraryPrefix()
```
Not needed anymore

trylek replied:
Deleted in 22nd commit.

trylek committed:
Addressing additional bits of JanK's PR feedback

Remove obsoleted methods GetLibraryPrefix and GetLibraryExtension;
Put installer pkgproj in sync with the updated clrjit naming;
Move _lastException to the other instance fields in CorInfoImpl;
Add TargetSpec mapping for all four CPU architectures.

jkotas commented on src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.cs:
```diff
@@ -110,14 +114,67 @@ private IntPtr AllocException(Exception ex)
+            string targetArchComponent = targetArchitecture switch
+            {
+                TargetArchitecture.X64 => "x64",
+                TargetArchitecture.ARM64 => "arm64",
```
Nit: I think you can fill all these in, even though we do not have them today, so that we do not have do it piecemeal later.

trylek replied:
Added in 22nd commit.

jkotas commented on src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.cs:
```diff
@@ -58,13 +58,17 @@ private enum ImageFileMachine
-        private extern static IntPtr getJit();
+        private static JitStartupDelegate s_jitStartup;
+        private static GetJitDelegate s_getJit;
+        private ExceptionDispatchInfo _lastException;
```
Nit: Keep this in the block with other instance vars?

trylek replied:
Moved in 22nd commit.

jkotas commented on src/installer/pkg/projects/netcoreapp/pkg/Microsoft.NETCore.App.Crossgen2.pkgproj:
```diff
@@ -42,14 +42,23 @@
       <Crossgen2File Include="$(CoreCLRCrossgen2Dir)System.CommandLine.dll" />
-      <Crossgen2File Include="$(CoreCLRCrossgen2Dir)$(LibraryFilePrefix)clrjitilc$(LibraryFileExtension)" />
       <Crossgen2File Include="$(CoreCLRCrossgen2Dir)$(LibraryFilePrefix)jitinterface$(LibraryFileExtension)" />
+      <Crossgen2File Include="$(CoreCLRCrossgen2Dir)$(TargetSpec)/$(LibraryFilePrefix)clrjitilc$(LibraryFileExtension)" />
```
Still using the old name (clrjitilc) ?

trylek replied:
Fixed in 22nd commit.

jkotas commented on src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunCodegenCompilationBuilder.cs:
```diff
@@ -9,6 +9,7 @@
 using ILCompiler.DependencyAnalysis;
 using ILCompiler.DependencyAnalysis.ReadyToRun;
 using ILCompiler.DependencyAnalysisFramework;
+using ILCompiler.PEWriter;
```
This using should not be needed.

trylek replied:
Deleted in 23rd commit.

jkotas commented on src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunCodegenCompilationBuilder.cs:
```diff
@@ -192,7 +193,7 @@ public override ICompilation ToCompilation()
                 corJitFlags.Add(CorJitFlag.CORJIT_FLAG_BBINSTR);
 
-            JitConfigProvider.Initialize(corJitFlags, _ryujitOptions, _jitPath);
+            JitConfigProvider.Initialize(_context.Target.OperatingSystem, _context.Target.Architecture, corJitFlags, _ryujitOptions, _jitPath);
```
It may be better to initialize JitConfigProvider with `_context.Target`, and let it pick the details it needs.

trylek replied:
Good point, changed in 23rd commit.

jkotas commented on src/installer/pkg/projects/netcoreapp/pkg/Microsoft.NETCore.App.Crossgen2.pkgproj:
```diff
@@ -42,14 +42,22 @@
+      <TargetSpec>$(TargetOSComponent)-$(TargetArchitecture)</TargetSpec>
+    </PropertyGroup>
+
+    <Error Condition="'$(TargetOSComponent)' == ''" Text="Unsupported target OS: '$(TargetOS)'" />
```
This can never happen now.

trylek replied:
Deleted in 23rd commit.

trylek committed:
Addressing additional JanK's feedback

- Use TargetDetails to pass the target OS / arch info;
- Remove obsoleted Error clause from the installer pkg script;
- Deleted unneeded using clauses from R2RCGCompilationBuilder.

AntonLapounov commented on src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.cs:
```diff
@@ -110,14 +115,44 @@ private IntPtr AllocException(Exception ex)
+                s_jitLibrary = NativeLibrary.Load(jitLibraryName, typeof(CorInfoImpl).Assembly, searchPath: null);
+            }
+
+            s_jitStartup = Marshal.GetDelegateForFunctionPointer<JitStartupDelegate>(NativeLibrary.GetExport(s_jitLibrary, "jitStartup"));
+            s_getJit = Marshal.GetDelegateForFunctionPointer<GetJitDelegate>(NativeLibrary.GetExport(s_jitLibrary, "getJit"));
```
What was wrong with the existed `NativeLibrary.SetDllImportResolver` approach?

trylek replied:
As JanK pointed out, with the changed approach to locating the JIT dll it no longer makes much sense - we either consume the stock clrjit next to Crossgen2 as identified by the targeting OS / CPU architecture, or explicitly load the custom one as specified in the JitPath command-line argument.

AntonLapounov replied:
> we either consume the stock clrjit next to Crossgen2 as identified by the targeting OS / CPU architecture, or explicitly load the custom one as specified in the JitPath command-line argument

And that logic could easily be placed in the `NativeLibrary.SetDllImportResolver` callback.  Am I missing anything?

MichalStrehovsky replied:
+1 - The SetDllImportResolver is cleaner and a lot fewer lines of code than juggling delegates. CorInfoImpl no longer needs to worry about paths. Are there any drawbacks I'm missing?

MichalStrehovsky replied:
(I think Jan's comment was that since you're juggling delegates, you don't need SetDllImportResolver, but Anton and I are questioning the juggling of delegates in the first place.)

trylek replied:
I have modified the change based on your feedback in 25th commit by basically reverting all remaining changes to CorInfoImpl and just handling the clrjit name translation logic in SetDllImportResolver.

trylek committed:
Fix clrjit path in Signing.props

AntonLapounov commented on src/coreclr/src/tools/crossgen2/crossgen2/crossgen2.csproj:
```diff
@@ -40,30 +40,68 @@
+      UseHardLinksIfPossible="true"
+      />
+
+    <Copy
+      SourceFiles="$(BinDir)\$(CrossHostArch)\$(JitInterfaceLibraryName)"
+      DestinationFiles="$(CrossPackageFolder)\$(JitInterfaceLibraryName)"
+      UseHardLinksIfPossible="true"
+      />
```
I think you can replace it with
```XML
<PackageFile Include="$(BinDir)\$(CrossHostArch)\$(JitInterfaceLibraryName)" />
```
added to the previous `<ItemGroup>`.

trylek replied:
Good point, fixed in 25th commit.

AntonLapounov commented on src/coreclr/src/tools/crossgen2/crossgen2/crossgen2.csproj:
```diff
@@ -40,30 +40,68 @@
+    </PropertyGroup>
+
+    <ItemGroup>
+      <PackageFile Include="$(BinDir)\crossgen2\*" Exclude="$(BinDir)\crossgen2\$(JitInterfaceLibraryName)" />
```
Should we also exclude `clrjit*`, which is an ARM64 binary?

trylek replied:
Fixed in 25th commit, thanks for pointing that out!

AntonLapounov commented:
> Targeting environment is represented by the second-level subfolder under the hosting folder

Please fix the PR and commit description before merging.  I have reviewed all the new commits — thanks for working on this!

trylek committed:
Put back SetDllImportResolver per Anton's and Michal's feedback

MichalStrehovsky reviewed (approved):
Thank you Tomas!

trylek committed:
Remove caching of clrjit handle per Michal's PR feedback
